### DISCUSSION TITLE
Android receives login failure when returning from background to foreground

### DISCUSSION THREAD

**Discussion by DavidWeng916** [pixiv/vroid-sdk-developers#86] [Q&A]
Hi, I would like to ask a question, when I publish to Android platform, I found that when my app goes back to the background and then back to the foreground, the VRoid SDK itself triggers a login failure, but I actually don't call any login related APIs.
How can I solve this problem? I am using VRoid SDK version 0.1.4

**Comment by nyamadan** [comment]
Him the server process seems to have succeeded because it came back from the background.
There may be an error inside SDK.

Could you try `ApiController.Logout`?

**Reply by DavidWeng916** [reply]
Hello @nyamadan, I'm not quite sure why I'm calling `ApiController.Logout`?
But I just tested it using the example project attached to the VRoid SDK (`Assets/VRoidSDK/Examples/CharacterModelExample/Scenes/Example.unity`) and it seems to have the same problem.

The following is my testing process
1. Logout first
2. Then login
3. Return the app to the background
4. Tap the app icon on the desktop again to return to the foreground

Then we can see in the program `ApiController.cs`
The function `LoginToVroidHub` in the program，Inside `_oauthClient.Login(onSuccess, onFailed) ` the onFailed will be called
I put log in onFailed this callback, please refer to the attached picture
<img width="958" alt="截圖 2023-06-06 下午3 17 50" src="https://github.com/pixiv/vroid-sdk-developers/assets/115971193/309bfcdc-7706-49a7-931d-70dcefea6228">

Whenever I go back to the foreground from the background onFailed is called, which causes some misjudgment.
`===== Error e:Pixiv.VroidSdk.Browser.RegisterCodeFailedException: getData is null
UnityEngine.Logger:Log(LogType, Object)
VRoidSDK.Examples.Core.Controller.<>c__DisplayClass16_0:<LoginToVroidHub>b__6(Exception)
Pixiv.VroidSdk.Browser.AndroidBrowser:OnApplicationFocus(Boolean)`

Then when I click on the menu above, an error message pops up.
![Screenshot_20230606-151401_Test_VRoidSDK_020](https://github.com/pixiv/vroid-sdk-developers/assets/115971193/283b9e30-7368-4a83-9328-51e0a70ba8d8)

For this test I am using VRoid SDK 0.2.0.
Because of this problem, I see another [problem](https://github.com/pixiv/vroid-sdk-developers/discussions/85) is about the error message language sometimes will be Japanese and sometimes will be English, and then I just found out that later sometimes there is also Chinese QQ

**Reply by nyamadan** [reply]
`ApiController.Logout` clears local data.
Sometimes this operation solves problem, but not in this case.

> `===== Error e:Pixiv.VroidSdk.Browser.RegisterCodeFailedException: getData is null UnityEngine.Logger:Log(LogType, Object) VRoidSDK.Examples.Core.Controller.<>c__DisplayClass16_0:<LoginToVroidHub>b__6(Exception) Pixiv.VroidSdk.Browser.AndroidBrowser:OnApplicationFocus(Boolean)`

Thank you, this log is useful for me.
It seems to fail to handle android browser.

I'll investigate.

**Reply by DavidWeng916** [reply]
Thank you very much 🙏

**Reply by DavidWeng916** [reply]
Hello, about this issue, are there any follow-up results of the survey?

**Reply by nyamadan** [reply]
We created an issue ticket in our task system and assigned an engineer.

It seems that `getIntent().getData()` is returning null.
https://stackoverflow.com/questions/28445262/getintent-getdata-null-for-unity-app
